### ISSUE TITLE
Replacing OpenTK

### ISSUE THREAD

**Issue by tomspilman** [MonoGame/MonoGame#4399]
This stems from a discussion started in another PR:

https://github.com/mono/MonoGame/pull/4397#issuecomment-169345797

It has been over 3 years since we initially investigated sticking with OpenTK in MonoGame.  The OpenTK team at that time promised fixes and renewed development of the library.

Since then there have been improvements, but we still suffer problems with OpenTK functionality and need per-platform work arounds at times.

Is it finally time to give up on OpenTK?

Do we proceed with our own custom binding for SDL2 and OpenGL?

What happens to platforms like iOS and Android where we depend on the Xamarin fork of OpenTK for OpenGL access?

**Comment by hach-que** [comment]
Is it not possible to P/Invoke native libraries on iOS / Android? I was under the impression that worked.

If it does, we could just P/Invoke OpenGL there too.

**Comment by tomspilman** [comment]
> Is it not possible to P/Invoke native libraries on iOS / Android? 

I think it is.

This would be a new path for the project.  Where in the past we depended on bindings provided by other open source projects or vendors like Xamarin.... we would be maintaining our own bindings.

It might be more work, but it puts us more in control over our destiny on these platforms.

Not this is sort of the same boat we are in on consoles where we generate our own bindings to native console APIs.

**Comment by dellis1972** [comment]
it is possible to PInvoke on both Android and iOS, I can certainly help out on the Android side.

**Comment by tomspilman** [comment]
Note that on iOS we may eventually want to use the new Metal API which I think is not available to C# in any way at the moment.  So there might be another binding to generate there at a later date.

**Comment by hach-que** [comment]
It also puts the project in a position where we could drop the dependency on Xamarin for iOS / Android if an alternative way of running IL on those platforms arose.

**Comment by tomspilman** [comment]
@hach-que - That would be possible too I guess.  Who knows where Microsoft's open source .NET stuff will lead in the coming years.

**Comment by tomspilman** [comment]
So here is something to consider.

On consoles the styles of APIs are all over the place... some are very simple C APIs, some are handle based, some use lots of macros, some take structures with dozens of parameters, and some are actual C++ class libraries with inlined functions and everything.  These all pose different challenges in writing bindings.

Some of the APIs we use from MonoGame and game code are called infrequently so any cost of making the C# to native call transition is cheap.  However some need to be called 100s of times a frame... typically rendering calls.  This can make the call overhead to bound APIs costly.

So what we have done with MonoGame on consoles:

 - Write a custom C++ class to wrap the console native API.
 - Use [CppSharp](https://github.com/mono/CppSharp) to generate C# bindings.

This way we end up with things like:
```
var nativeTexture = new Some.Console.Graphics.Texture(format, width, height, mips);
```
We basically hide the internals of the native console API behind a C# binding class that just exposes the features MonoGame needs.  Internally that construction above may make dozens of native API calls internally using all sorts of different API standards... but to C# it is one clean looking API.

This technique also allows us to expose additional console native APIs in nice C# libraries that developers can easily consume.  This comes up a lot as consoles have specialized features outside of what MonoGame provides that developers need to access as well.

The downside to all this is this extra layer of C++ code to maintain that can contain the typical C++ bugs.

I bring this all up so that we can consider this as another path beyond simple pinvokes.

**Comment by hach-que** [comment]
In that context I should mention that Protobuild supports generating
cross-platform C++ projects and allows you to reference them from C#
projects (and it handles configuring the projects so the native DLLs get
copied out to the correct place).

**Comment by KonajuGames** [comment]
Xamarin do expose all the Metal APIs to C#.

**Comment by tomspilman** [comment]
> Xamarin do expose all the Metal APIs to C#.

Great to know!

**Comment by flibitijibibo** [comment]
If y'all want some single-file OpenGL bindings that are easy to statically link, FNA's bindings are here:

https://github.com/FNA-XNA/FNA/blob/master/src/Graphics/OpenGLDevice_GL.cs

Note the LoadGLEntryPoints func - by loading the entry points by hand it gives you a LOT of room to customize for various GL targets, and also gives you the ability to throw NoSuitableGraphicsDeviceExceptions when the user doesn't have the necessary entry points.

There's also SDL2# and OpenAL# which are easy to statically link.

**Comment by tomspilman** [comment]
@flibitijibibo - Thanks we will look, but will likely roll our own so we can consider mobile and console concerns.

**Comment by xanather** [comment]
On Desktop may I suggest interoping with GLFW https://github.com/glfw/glfw. I do this for my own cross-plat 2D engine. Handles windowing and input very well across different platforms. OpenGL is still used directly/natively.

**Comment by KonajuGames** [comment]
Switching one third-party dependency for another is not really moving forward.  And GLFW needs another dependency that is the C# bindings.  The intention was to remove that dependency.

**Comment by cra0zy** [comment]
I vote we replace OpenTK with SDL2 on DesktopGL platform, there are some benefits for getting MG working on Web :)

**Comment by cra0zy** [comment]
We can replace OpenTK for DesktopGL in bursts:
1) Joystick and Gamepad
2) Window and Mouse/Keyboard
3) Graphics

Now as for should we use CppSharp for DesktopGL

Pros that I see:
 - easier setting up of bindings, since we can use native functions and enums without defining them
 - some performance gain

Cons that I see:
 - adds c++ code to c# reposetory
 - needs to compile for each OS
   - you will need a Linux buildbox then
 - needs an extra library when exporting compared to not using CppSharp
   - that library will probably have to be setuped as "copy to output folder"

I am ok with whatever you choose

**Comment by tomspilman** [comment]
> •adds c++ code to c# repository

I really would hate to do that....  the more pure C# this project can be the better.  This is really one of the only reasons I hesitate at the CppSharp route.

Really your list has more cons than pros...  maybe we should just go with manually generated pinvokes?

Ideally this would all be in one C# file only included in GL platforms including mobile.  We only expose the OpenGL API bits that we need.  We can maybe then add some helper functions over it to hide GL/GLES differences.

Thoughts?

**Comment by cra0zy** [comment]
> Ideally this would all be in one C# file only included in GL platforms including mobile. We only expose the OpenGL API bits that we need. We can maybe then add some helper functions over it to hide GL/GLES differences.

I like it.

**Comment by dellis1972** [comment]
>Ideally this would all be in one C# file only included in GL platforms including mobile. We only expose the OpenGL API bits that we need. We can maybe then add some helper functions over it to hide GL/GLES differences.

I'd rather go down this route. I think.

**Comment by tomspilman** [comment]
@cra0zy - Want to take a first stab at Joystick and Gamepad support via SDL2 using this technique?

I setup a new branch from 'develop' for this...  

  https://github.com/mono/MonoGame/tree/replace_opentk

**Comment by cra0zy** [comment]
> Want to take a first stab at Joystick and Gamepad support via SDL2 using this technique?

I started working on that 5 minutes before you made a comment :D

> Just make feature branches from it and submit PRs back to it so others can test/contribute as well.

Will do.

**Comment by WardBenjamin** [comment]
@cra0zy We in fact can! I got permission from him to use it for my own framework and he mentioned that. If you check it out, there's example code in FNA then.

**Comment by flibitijibibo** [comment]
I'm going to be shifting some things around quite a bit in the SDL2 folder over the next couple days, so be aware of that if you want to grab references from there. (But yes, please do use whatever you like from either project, whatever makes the transition away from OpenTK easier.)

**Comment by tomspilman** [comment]
> can we use @flibitijibibo SDL2-CS source files(this way no external dll) and modify it as we go along?

I'd rather see us create and maintain our own wrapper like I mentioned above.

**Comment by cra0zy** [comment]
> I'm going to be shifting some things around quite a bit in the SDL2 folder over the next couple days, so be aware of that if you want to grab references from there. (But yes, please do use whatever you like from either project, whatever makes the transition away from OpenTK easier.)

YOU ARE AWESOMEEE!!!!

> I'd rather see us create and maintain our own wrapper like I mentioned above.

Well I don't really see a point in creating something that's already created... especially in this case since @flibitijibibo created bindings with mostly functions for FNA, and MonoGame is probably gonna need the exact same functions.

As for updating bindings, as I said, we would use that as base, and then just update stuff as we need it.

Also: https://github.com/flibitijibibo/SDL2-CS/blob/master/src/SDL2.cs#L3781
That's painful...

<b>Forgot to mention in comment above about stuff to replace</b>
We will also need to replace OpenAL stuff as well... well I will be working on this as 3rd thing on the list, someone else will be able to work on OpenGL in parallel.

**Comment by cra0zy** [comment]
Side note, on the topic of documentation and writing bindings, Gtk has the best documentation I have ever seen, just look at this: https://developer.gnome.org/gtk3/stable/GtkTreeView.html

**Comment by cra0zy** [comment]
@flibitijibibo In SDL2 wrapper, how did you calculate padding variable in structs? ie. https://github.com/flibitijibibo/SDL2-CS/blob/master/src/SDL2.cs#L3685

**Comment by tomspilman** [comment]
> how did you calculate padding variable in structs

Usually via a compiler option or #pragma the padding for types is defined in C++.  Often the default padding is 4 bytes which is what I would assume SDL is using.  So any types less than 4 bytes get padded out to 4 bytes.  This padding could occur between members too, so you have to look at things carefully.

**Comment by flibitijibibo** [comment]
That's actually straight out of the headers:

https://hg.libsdl.org/SDL/file/d710a7a66aa5/include/SDL_events.h#l373

SDL2 does not enforce any kind of struct padding, so we don't either. The only time you should expect padding that _isn't_ straight from the headers is when a CLR is found to not be accurate for some reason... which can happen, believe it or not:

https://github.com/flibitijibibo/TheoraPlay-CS/blob/master/TheoraPlay.cs#L44
https://github.com/flibitijibibo/TheoraPlay-CS/blob/master/TheoraPlay.cs#L155

**Comment by dellis1972** [comment]
Another thing we should think about. There are a new breed of mobile GPU's which support Full OpenGL 4, there is at least one device that I know of that uses Android and have Full OpenGL. We should be able to switch over to Full GL on and Android device at runtime. So when we design the wrapper we need to take this into account. 

Currently you can do stuff like this on android

```
SupportsFullGL = Egl.BindAPI (Egl.RenderApi.GL);
 if (!SupportsFullGL)
     Egl.BindAPI (Egl.RenderApi.ES);
```
Which will bind the GL or GLES API , you then need to load the correct entry points. But then you end up with two code paths when calling the API's because they are slightly different.
So anything we can do to make the GL entry points the same across Desktop/Android/iOS will make this easier.

**Comment by cra0zy** [comment]
What are we gonna do about Angle platform? It's using OpenTK.

**Comment by tomspilman** [comment]
> What are we gonna do about Angle platform? It's using OpenTK.

Angle mimics GL APIs... so in theory we could use the same bindings against the angle DLL.  Might be able to setup a trick where we only have one implementation of the bindings and overload the DLL filename string passed to the `DllImportAttribute`.

**Comment by cra0zy** [comment]
I submitted #4428 so that multiple people can work on different parts of this.

My next goal is setting up GamePad.

Available stuff that shouldn't cross paths are:
 - Song - should get replaced with sdl_mixer
 - SoundEffect - needs our own OpenAL bindings
 - Window / Platform
 - New GL bindings

If you want to start working on something say it in here so that there isn't double work done.

**Comment by dellis1972** [comment]
I'll grab GL bindings as I've probably had the most experience across desktop and mobile platforms

-----Original Message-----

**Comment by tomspilman** [comment]
> I'll grab GL bindings

Fantastic!

**Comment by cra0zy** [comment]
> I'll grab GL bindings as I've probably had the most experience across desktop and mobile platforms

Here is a weird thought, while you are setting up the common path for Desktop and Mobile, could you add it to Web as well?

You don't have to setup any bindings, but instead call it as if you were calling javascript(which you are).

Webgl function:
https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Web/WebHelper.cs#L9

It's getting set in here:
https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Web/WebGameWindow.cs#L37

An example of calling:
https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Graphics/GraphicsDevice.Web.cs#L30

List of webgl functions:
https://www.khronos.org/files/webgl/webgl-reference-card-1_0.pdf

**Comment by cra0zy** [comment]
For an hour I've been trying to setup gamepad vibrations, I even went over @flibitijibibo code trying to figure it out, and only now do I realize that the code from an hour ago was ok, and all I needed was SDL_INIT_HAPTIC flag during int....

**Comment by cra0zy** [comment]
Does anyone know why `SDL_GetError()` is causing stack trace?

**Comment by flibitijibibo** [comment]
The marshaler might be freeing a const char*, have a look at SDL2#'s string marshaler to see if that fixes it.

**Comment by tomspilman** [comment]
> The marshaler might be freeing a const char*, 

I suggest changing your function pinvoke signature to return a `IntPtr` and see if it still crashes.  If it does then it has nothing to do with any marshalling.

**Comment by cra0zy** [comment]
> The marshaler might be freeing a const char*, have a look at SDL2#'s string marshaler to see if that fixes it.

-

> I suggest changing your function pinvoke signature to return a IntPtr and see if it still crashes. If it does then it has nothing to do with any marshalling.

@flibitijibibo was right.

Thank you both for the info.

I ended up adding `GetSDLString` method to just convert it from pointer.

**Comment by tomspilman** [comment]
> I ended up adding  SDL_GetString  method to just convert it from pointer.

How were you marshaling it before?  Maybe you had it wrong.

In these cases where I have to do some extra work to pass or return parameters...  I make the pinvoke method private then add a friendly public method that hides the extra conversion work.

**Comment by flibitijibibo** [comment]
Here's SDL2#'s marshaling pull request, if you want some additional context as to what needs the conversion work:

https://github.com/flibitijibibo/SDL2-CS/pull/22

The non-custom marshaling worked for us, but UTF8 is nice to have! If .NET catches up then this shouldn't be necessary.

**Comment by cra0zy** [comment]
> How were you marshaling it before? Maybe you had it wrong.

I didn't... With Gtk I didn't have to do it, and SDL required it and I didn't notice,

> In these cases where I have to do some extra work to pass or return parameters... I make the pinvoke method private then add a friendly public method that hides the extra conversion work.

https://github.com/mono/MonoGame/commit/2986b861f10184e1e7013431958d016d1e6855f0

> Here's SDL2#'s marshaling pull request, if you want some additional context as to what needs the conversion work:
> 
> flibitijibibo/SDL2-CS#22
> 
> The non-custom marshaling worked for us, but UTF8 is nice to have! If .NET catches up then this shouldn't be necessary.

@tomspilman @flibitijibibo If both of you are fine with this I don't mind replacing my code with that one.

**Comment by tomspilman** [comment]
@cra0zy 

What I meant was more like this:

```
private unsafe static string StringFromUtf8(IntPtr utf8String)
{
    if (utf8String == IntPtr.Zero)
        return "";

    var ptr = (byte*)utf8String;
    while (*ptr != 0)
        ptr++; 

    var bytes = new byte[ptr - (byte*)utf8String];
    Marshal.Copy(utf8String, bytes, 0, bytes.Length);

    return Encoding.UTF8.GetString(bytes);
}

[DllImport(nativeLibName, CallingConvention = CallingConvention.Cdecl)]
private static extern IntPtr __SDL_GameControllerName(IntPtr gamecontroller);

public static string SDL_GameControllerName(IntPtr gamecontroller)
{
    return StringFromUtf8(__SDL_GameControllerName(gamecontroller));
}
```
I do however hate that we are generating an extra garbage byte array per marshaled string there.  Don't know if we can pool that without a lot of thread safety overhead.

> I don't mind replacing my code with that one

The problem I have with using the marshalling helpers via attributes is that it all goes thru an interface making several slower than regular function calls to convert a single string. 

On consoles we avoid all the `MarshalAs` stuff for performance.

**Comment by cra0zy** [comment]
> What I meant was more like this:

I did understand what you meant, I just wanted to show you what I did, plus ask about the attribute way before making changes.

> [DllImport(nativeLibName, CallingConvention = CallingConvention.Cdecl)]
> private static extern IntPtr __SDL_GameControllerName(IntPtr gamecontroller);
> 
> public static string SDL_GameControllerName(IntPtr gamecontroller)
> {
>     return StringFromUtf8(__SDL_GameControllerName(gamecontroller));
> }

You are missing an EntryPoint :P

> The problem I have with using the marshalling helpers via attributes is that it all goes thru an interface making several slower than regular function calls to convert a single string.

Didn't know that O o.

> I do however hate that we are generating an extra garbage byte array per marshaled string there. Don't know if we can pool that without a lot of thread safety overhead.

The only wat that is possible is if we could convert pointer to array without making a copy of it, or set up our own UTF8 encoder... first one seems impossible from c#, second one is way too much work.

**Comment by cra0zy** [comment]
> Window / Platform

I am taking this.

**Comment by cra0zy** [comment]
@tomspilman Can you setup a branch in MonoGame.Dependencies, I need to modify opentk.dll.config so I can force opentk to use sdl window.

**Comment by tomspilman** [comment]
> Can you setup a branch in MonoGame.Dependencies,

How would that end up working if and when we decide to merge this experiment into develop?

We would have to first merge this new `replace_opentk` MonoGame.Dependencies branch into develop in MonoGame.Dependencies.

Then when merging `replace_opentk` into `develop` here in the main repo we would have to change the submodule to point to the new revision.

This sound correct?  @KonajuGames @dellis1972 ?

**Comment by KonajuGames** [comment]
That does sound like the correct way to do the work across multiple repos.

**Comment by elisee** [comment]
> @tomspilman @flibitijibibo - Thanks we will look, but will likely roll our own so we can consider mobile and console concerns.

Have you read @flibitijibibo's [recent post about FNA's new platform abstraction layer](https://plus.google.com/+flibitijibibo/posts/L7dFrURPPEz)? It paves the way for mobile and console support (although flibit is not interested in working on further platform himself). Just thought it might be worth looking more seriously into what can be reused / shared here.

**Comment by flibitijibibo** [comment]
The FNAPlatform stuff doesn't help too much since it basically threw out GamePlatform, though the static SDL2_FNAPlatform methods did definitely clean those files up a bit. You could contort them back into a GamePlatform pretty easily since the rewrite just happened last week, but the actual changes made would affect a LOT more here than what it affected for us.

**Comment by tomspilman** [comment]
I don't see us using the FNA platform stuff.  Avoiding interfaces for abstraction is a deliberate choice in XNA and MonoGame as they are the slowest of method calls.

**Comment by Frassle** [comment]
Just to give some fuel to this project, I'm [notifying](https://github.com/opentk/opentk/commit/2e51fe019771f4c70b15bb25b3a560f712a45fbf) that OpenTK is probably dead.

I seem to be the only one still keeping it hobbling along and I just don't have the access rights or position to keep it going.

**Comment by dellis1972** [comment]
@Frassle thanks for the heads up... 

@tomspilman can we fork OpenTK to MonoGame (I don't have permissions) so we can at least get fixes in until we can replace it entirely. @mrhelmut we'll need to pull the PR's over to the new fork asap.

**Comment by tomspilman** [comment]
> can we fork OpenTK to MonoGame 

Specifically a fork of https://github.com/opentk/opentk ?  Correct?

**Comment by dellis1972** [comment]
yes please

**Comment by tomspilman** [comment]
@dellis1972  - Done...  https://github.com/MonoGame/opentk

@cra0zy - Done...  https://github.com/MonoGame/MonoGame.Dependencies/tree/replace_opentk

**Comment by dellis1972** [comment]
Awesome

**Comment by mrhelmut** [comment]
Great news. I was about to ask if I could push an updated binary to the dependency repo, but a fork will make things easier while this issue is being worked on. I'll push the changes tomorrow.

**Comment by mrhelmut** [comment]
FYI MonoGame/opentk#1 MonoGame/opentk#2

Once merged, would you need to push the assembly to MonoGame/MonoGame.Dependencies ?

**Comment by tomspilman** [comment]
> Once merged

Those are now merged.


> would you need to push the assembly to MonoGame/MonoGame.Dependencies ?

I would first be sure we don't need to do any other immediate fixes to OpenTK...  @cra0zy ?

After that someone needs to build our fork of OpenTK and submit a PR to the dependencies repo and submit a PR updating the MonoGame submodule reference to the dependencies.

**Comment by dellis1972** [comment]
@tomspilman can u grant me merge rights on the opentk repo plz I was going to merge earlier but couldn't doh :)

-----Original Message-----

**Comment by tomspilman** [comment]
@dellis1972 - Done!

**Comment by dellis1972** [comment]
Thanks :)

-----Original Message-----

**Comment by cra0zy** [comment]
We need one more thing in out opentk repo, just a simple public variable to see if it's running SDL or native backend, #4398 #4436

Alternatively we can remove centering code.

**Comment by mrhelmut** [comment]
Can't it be determined with ```OpenTK.Configuration.RunningOnSdl2```?

**Comment by cra0zy** [comment]
> Can't it be determined with OpenTK.Configuration.RunningOnSdl2?

I didn't know about it's existence...

**Comment by cra0zy** [comment]
> Ideally this would all be in one C# file only included in GL platforms including mobile. We only expose the OpenGL API bits that we need. We can maybe then add some helper functions over it to hide GL/GLES differences.

Should we do something like that for OpenAL as well?

**Comment by dellis1972** [comment]
> Should we do something like that for OpenAL as well?

probably not sure there are too many differences in that api though.

**Comment by cra0zy** [comment]
> probably not sure there are too many differences in that api though.

I don't know how much API's are different, It's more that some functions act differently depending on the platform.

**Comment by dellis1972** [comment]
OK, I started work on a low level OpenGL binding 

https://github.com/infinitespace-studios/OpenGLSharp

Wanted feedback on the technique before going any further :) Tested on iOS and Android, not SDL yet. 
Note the *MonoNativeFunctionWrapper* is required for iOS in order to use *Marshal.GetDelegateForFunctionPointer*. 

The main reasoning for going down this route is a) it works better with SDL and more importantly b) new mobile GPU's support FULL OpenGL and with Android you can get devices that support both. So we might as well use it if its there :)

**Comment by cra0zy** [comment]
Do you plan on making them a part of MonoGame repo, or do you plan on having a separate repo/dll file?

**Comment by dellis1972** [comment]
Not thought abt it really,it can probably live in the MG repo :) as it will really only support the min that MG needs and nothing else.

-----Original Message-----

**Comment by tomspilman** [comment]
> Not thought abt it really,it can probably live in the MG repo 

Yeah I think that would be better.  Makes it easier to update and version along with MG.

**Comment by dellis1972** [comment]
Ok cool, once we agree its a good direction then I'll start filling it in :)

-----Original Message-----

**Comment by cra0zy** [comment]
I am just gonna say what's on my mind right now... webgl... I only have 4 bytes of memory...

**Comment by KonajuGames** [comment]
You might have to enlighten some of us what that means in regards to the
bindings. I have no idea what WebGL needs in terms of bindings. I'm
assuming it needs bindings. If so, are they different to the other
platforms?

**Comment by cra0zy** [comment]
> You might have to enlighten some of us what that means in regards to the
> bindings. I have no idea what WebGL needs in terms of bindings. I'm
> assuming it needs bindings. If so, are they different to the other
> platforms?

That's the funny thing.. it doesn't need bindings O o

https://github.com/mono/MonoGame/issues/4399#issuecomment-171780119

**Comment by KonajuGames** [comment]
We can simply not include the bindings in the WebGL project then. Is that
right?

**Comment by cra0zy** [comment]
> We can simply not include the bindings in the WebGL project then. Is that
> right?

We don't have any WebGL bindings set at all... I just thought it would be nice to have an interface that works for WebGL as well.

**Comment by dellis1972** [comment]
Why oh why do they have to use "clear" in stead of "Clear" in webgl ug :/

**Comment by cra0zy** [comment]
I wanted to talk about fullscreen, I think we should use "SDL_WINDOW_FULLSCREEN_DESKTOP" instead of "SDL_WINDOW_FULLSCREEN". What't the difference? With the first flag we aren't changing the resolution, but instead we are just making the window fullscreen, this means that we need to modify the viewport to set the correct size.

On Linux, games usually don't change the resolution of the screen but instead just change the viewport.

Advantages for SDL_WINDOW_FULLSCREEN_DESKTOP:
 - Faster switching between window/fullscreen
 - Faster changing of size when in fullscreen
 - You can use any resolution while fullscreen
 - It won't damage multimonitor setup

Disadvantages:
 - none that I can see...

We could also maybe setup a flag to let user decide on which type of fullscreen they want.

**Comment by cra0zy** [comment]
> Why oh why do they have to use "clear" in stead of "Clear" in webgl ug :/

Copyright :laughing: 

Why is that causing you trouble?

**Comment by KonajuGames** [comment]
> We could also maybe setup a flag to let user decide on which type of fullscreen they want.

https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/GraphicsDeviceManager.cs#L488

**Comment by dellis1972** [comment]
because web GL is dynamic, we can just get the "GL" using

```
GL = glcanvas.getContext("webgl");
```

because our bindings are in a class called GL.. that might work nicely for sharing code. however because 
of the "clear" "Clear" issue that means ALL the bindings for the other platforms will need to use the same naming conventions as webgl. which is not really the C# way :). I support we could have wrappers that forward the calls to webgl that would fix it just a bit more code for webgl.. I could probably auto generate those wrappers.

**Comment by KonajuGames** [comment]
I don't want to compromise C# naming for one platform.  If/when this SDL2 change is done, I want to refactor the SDL2 bindings because they violate every C# naming standard (one of the reasons I didn't like them when they were first proposed a long time back).

**Comment by dellis1972** [comment]
@KonajuGames totally agree :+1: I'll find a way to support both :)

**Comment by cra0zy** [comment]
> https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/GraphicsDeviceManager.cs#L488

I had no idea we had that flag... O o AWESOME

> I don't want to compromise C# naming for one platform. If/when this SDL2 change is done, I want to refactor the SDL2 bindings because they violate every C# naming standard (one of the reasons I didn't like them when they were first proposed a long time back).

Can you give me an example of what would you like changed in? https://github.com/cra0zy/MonoGame/blob/sdl2/MonoGame.Framework/SDL/SDL2.cs

**Comment by KonajuGames** [comment]
> Can you give me an example of what would you like changed in?

- Nothing in all caps with underscores.
- Remove all SDL_ prefixes.  It's already in a SDL namespace.
- Constants to enums.

```
    public const uint   SDL_INIT_VIDEO          = 0x00000020;
    public const int    SDL_INIT_JOYSTICK       = 0x00000200;
    public const uint   SDL_INIT_HAPTIC         = 0x00001000;
    public const uint   SDL_INIT_GAMECONTROLLER = 0x00002000;
```
to
```
    [Flags]
    public enum Init
    {
        Video = 0x0020,
        Joystick = 0x0200,
        Haptic = 0x1000,
        GameController = 0x2000
    }
```
and
```
    public enum SDL_GameControllerButton
    {
        SDL_CONTROLLER_BUTTON_INVALID = -1,
        SDL_CONTROLLER_BUTTON_A,
        SDL_CONTROLLER_BUTTON_B,
        SDL_CONTROLLER_BUTTON_X,
```
to
```
    public enum ControllerButton
    {
        Invalid = -1,
        A,
        B,
        X,
```

- Wrap the Window functions into a Window class.
```
    public class Window : IDisposable
    {
        private IntPtr _handle;

        public Window(string title, int x, int y, int w, int h, WindowFlags flags);
        public Rect Bounds { get; set; }
        public bool Bordered { get; set; }
```
- Same with Joystick, GameController, Haptic, etc.

**Comment by KonajuGames** [comment]
This is something that can be refactored afterwards.  It doesn't have to be done now.

**Comment by cra0zy** [comment]
> Wrap the Window functions into a Window class.
> Same with Joystick, GameController, Haptic, etc

That seems a bit pointless since we would be doing Joystick > SdlJoystick > SDL_Joystick

We would be creating a middle layer which wouldn't have any use. What's more we would have a bit of performance loss, plus it also creates harder maintenance and much more code.

**Comment by tomspilman** [comment]
> Wrap the Window functions into a Window class

Yeah... lets not do that.  If we want the style more C# like the most we should do is change the pinvoke, enum, and struct names.  

```
[DllImport(nativeLibName, CallingConvention = CallingConvention.Cdecl), EntryPoint = "SDL_CreateWindow"] 
public static extern IntPtr CreateWindow(string title, int x, int y, int w, int h, SDL_WindowFlags flags);
```

That has zero performance impact and should solve most of your style itch.  It should be no worse than any OpenGL bindings then.

**Comment by cra0zy** [comment]
> Wrap the Window functions into a Window class
> Yeah... let's not do that. If we want the style more C# like the most we should do is change the pinvoke, enum, and struct names.
> 
> [DllImport(nativeLibName, CallingConvention = CallingConvention.Cdecl), EntryPoint = "SDL_CreateWindow"] 
> public static extern IntPtr CreateWindow(string title, int x, int y, int w, int h, SDL_WindowFlags flags);
> That has zero performance impact and should solve most of your style itch. It should be no worse than any OpenGL bindings then.

I can get behind that. I'll do it in #4428

**Comment by KonajuGames** [comment]
The naming is the primary aim. The classes are secondary, and I'm fine with
them not being done.

**Comment by cra0zy** [comment]
@KonajuGames @tomspilman What do you think? Want me to change something? https://github.com/cra0zy/MonoGame/blob/sdl2/MonoGame.Framework/SDL/SDL2.cs

**Comment by cra0zy** [comment]
Also do you want me to use `ushort` or `UInt16'?

**Comment by tomspilman** [comment]
> Also do you want me to use `ushort` or `UInt16`?

The keyword types like `ushort` are better.

**Comment by tomspilman** [comment]
> What do you think? Want me to change something? 

Looks good overall to me.  Not alot to say... it is simple and sticking to just what we use and nothing more.

I am a little unsure about the nested static classes...  `SDL.Window.Create()` instead of `SDL.CreateWindow()`.

Speaking of those nested static classes you can make them `public`.  The `SDL` class is already internal making them unreachable to anyone outside the assembly.

Also should we call this class `SDL2`?  Would it benefit MonoGame core devs to be reminded on every call that this is SDL2 and not SDL?

Finally is it worth sharing a static scratch byte array and locking it on use in `SDL.GetString()`.  This would avoid garbage at the cost of a lock.  Although it doesn't seem like that function is used much to begin with.

**Comment by cra0zy** [comment]
> I am a little unsure about the nested static classes... SDL.Window.Create() instead of SDL.CreateWindow().

At first I wasn't going to use nested classes, but autocomplete was getting filled with too many results, this way it groups it and makes it a bit easier to use IMO.

> Speaking of those nested static classes you can make them public. The SDL class is already internal making them unreachable to anyone outside the assembly.

Sure.

> Also should we call this class SDL2? Would it benefit MonoGame core devs to be reminded on every call that this is SDL2 and not SDL?

The file itself is called SDL2, but the API is just SDL(I never saw anyone change the name of the API with the version of the library, take Gtk for example, Gtk 1, 2 and 3 all use class name Gtk).

> Finally is it worth sharing a static scratch byte array and locking it on use in SDL.GetString(). This would avoid garbage at the cost of a lock. Although it doesn't seem like that function is used much to begin with.

We don't know the size of the information :(

**Comment by tomspilman** [comment]
> but the API is just SDL

That is a good point... the SDL team didn't change their `SDL_` wart.

> We don't know the size of the information 

It is pretty simple... you just resize the byte array if it is smaller than you need for a particular string.

But it probably isn't important for how infrequent string conversion should be in the API as it is now.

**Comment by cra0zy** [comment]
> It is pretty simple... you just resize the byte array if it is smaller than you need for a particular string.

That would only help if the new string is the same size as the old one. In all other cases it would cause a new memory allocation.

> But it probably isn't important for how infrequent string conversion should be in the API as it is now.

I agree, tho this problem really bugs me for some reason.

**Comment by cra0zy** [comment]
I just remembered, we didn't really talk about replacing some OpenTK stuff on Android and iOS. 

So, SDL does support Android and iOS, we could use it I guess?

**Comment by dellis1972** [comment]
@cra0zy it does, but from what I've seen its a bit of a hack. I plan to use the new OpenGL wrapper for iOS and Android. But we'll keep the windowing systems we have atm.

**Comment by dellis1972** [comment]
@cra0zy SDL2 is just for cross platform Desktop.

**Comment by cra0zy** [comment]
Are we using OpenTK for only GL stuff on Android/iOS?

**Comment by dellis1972** [comment]
@cra0zy atm DesktopGL, Android and iOS use OpenTK.. post this work
DesktopGL -> SDL2 + new OpenGL Bindings
Android/iOS -> new OpenGL Bindings + current native views


if you look at the stuff I'm working on for the new OpenGL bindings (https://github.com/infinitespace-studios/OpenGLSharp) you can see I plan to support android and iOS. But we don't need to use SDL2 on those platforms.. it adds a native dependency that we don't really need .

**Comment by cra0zy** [comment]
> @cra0zy atm DesktopGL, Android and iOS use OpenTK.. post this work
> DesktopGL -> SDL2 + new OpenGL Bindings
> Android/iOS -> new OpenGL Bindings + current native views
> 
> if you look at the stuff I'm working on for the new OpenGL bindings (https://github.com/infinitespace-studios/OpenGLSharp) you can see I plan to support android and iOS. But we don't need to use SDL2 on those platforms.. it adds a native dependency that we don't really need .

You misunderstood my intentions. I thought we were using OpenTK for getting some stuff, like keyboard input or touch input on Android/iOS, but now I know we aren't.

If we were using OpenTK for it, just using SDL would be a easy(lazy) way of fixing it, but since we aren't, there is no point in using SDL.

**Comment by tomspilman** [comment]
> If we were using OpenTK for it

To be clear... the goal is a full purge of OpenTK.  We don't plan to use it on any platform for any reason.

Xamarn's Android and iOS C# runtimes include their modified copies of OpenTK just for the OpenGL bindings.  While they actually work relatively well now...  in the past we've had trouble with inconsistencies with naming, ref vs out, namespaces, etc.  So instead we want to use our new native bindings which we can ensure is consistent with bindings on our other platforms.

**Comment by tanis2000** [comment]
> If we were using OpenTK for it, just using SDL would be a easy(lazy) way of fixing it, but since we aren't, there is no point in using SDL.

why aren't you using SDL for iOS and Android as well without using yet another GL binding? Or why aren't you using the new OpenGL bindings for desktop as well then? And use SDL just for window/input/etc...?

I mean, wouldn't it be better to stick to a single API for all the platforms?

**Comment by dellis1972** [comment]
We are using the new "simpler" GL binding on all the OpenGL platforms, iOS,
Android and Desktops (Linux, Mac, Windows). We plan to use SDL2 for
Window/Input on Desktop but keep with the native implementations of
GamePad/Input and Window on iOS and Android mostly because with those
implementations its easier to integrate with Gamer services etc. Plus at
the moment there is no need to ship any native .dylibs on iOS and hopefully
with the OpenAL stuff @KonajuGames <https://github.com/KonajuGames> is
working on we can drop penal_soft on android too and have a purely C#
implementation (like we do for DX).

To be honest if we had a decent cross platform windowing system that was
pure C# and didn't rely on a native dll/so/dylib (like SDL2 does) we'd
probably use that.. OpenTK tried... but fell short.

**Comment by KonajuGames** [comment]
Yes, I want to drop penal_soft. ;)

**Comment by dellis1972** [comment]
sometimes I hate the spell checker... some times I don't :)

**Comment by grbd** [comment]
Hi I've recently been looking into a cross platform way to embed a Monogame window / game into a user control for GtkSharp when I stumbled across this thread. Most of the online examples seem to involve grabbing a hold of the opentk window.
If the plan is to move away from opentk could I please request that some consideration be given towards the embedding of GraphicsDevice into an existing form for gtksharp / winforms / wpf etc
I really like the idea of a 3D .Net library that I can use on ether a desktop or an rpi, given that the upcoming 4.5 kernel will have 3d acceleration support for it.

**Comment by hach-que** [comment]
I have a patched version of MonoGame that supports embedding into GTK#, although it wasn't accepted upstream.  Here is the patched version: https://github.com/RedpointGames/MonoGame (it is constantly rebased on the latest version of MonoGame)

**Comment by grbd** [comment]
Thanks for the link that's very usefull (I spotted the pull request but wasn't sure if the fork was uptodate), I'm hopeful with the move away from opentk towards sdl that intergration will become easier / part of mainline eventually

**Comment by mrhelmut** [comment]
I'm facing more and more OpenTK related issues (the most important being the DPI scaling).

What can I do to help?

**Comment by cra0zy** [comment]
Currentlly in works:
 - GL bindings (@dellis1972) repo is at https://github.com/infinitespace-studios/OpenGLSharp if you want to help out with it.

Current list of open tasks:
 - AL bindings.
 - Test replace_opentk branch on Mac and Windows.

Maybe tasks:
 - Make Song use SDL_Mixer.

Later tasks (after OpenTK has been removed as a dependency):
 - Update project templates.
 - Fix up game exit if needed.

**Comment by smack0007** [comment]
Maybe this can be of some help to you guys:

https://github.com/smack0007/GLDotNet

I'm generating an OpenGL binding completely from the Xml spec files. My generator project could be forked to fit your needs if nothing else.

**Comment by cra0zy** [comment]
So, in order to make sure that everything is nice and up to date I thought that it would be better to use my fork for SDL2 development.

If anyone here has a problem with that speak now or forever be silent.

If everyone is OK with that, submit PRs to https://github.com/cra0zy/MonoGame/tree/SDL2 for any change you wish to do for it.

**Comment by cra0zy** [comment]
> Make Song use SDL_Mixer.

I forgot why SDL_Mixer was bad... it's bad because there is no way to get song position or song duration.

**Comment by mrhelmut** [comment]
Maybe vorbisfile/libvorbis is worth looking at (I haven't checked that yet, but that came to my mind). The ```OggStreamer``` using NVorbis looks stable now, though.

And I still haven't started working on the AL part, maybe by the end of this week.

**Comment by cra0zy** [comment]
I'm fine with us using NVorbis.

**Comment by dellis1972** [comment]
@mrhelmut Part of my work is for OpenGL has been doing the same for OpenAL. since I removed OpenTK completely nothing would compile. I'll try and push my work in progress this week

**Comment by mrhelmut** [comment]
@dellis1972 I'll gladly help.

**Comment by cra0zy** [comment]
@flibitijibibo Have you come across this weird bug by any chance https://github.com/cra0zy/MonoGame/blob/SDL2/MonoGame.Framework/SDL/SDLGameWindow.cs#L224 ?

**Comment by flibitijibibo** [comment]
I've not seen it, but I've also not come across a lot of resizable windows. You may want to look at how window position is stored, as it's a value kept internally by the window structure rather than something returned by the window server every time you call GetPosition. Though, for what it's worth, I believe XNA just centers the window on device changes anyway... that's what I ended up doing in FNA, so I'm sorta sidestepping this.

**Comment by cra0zy** [comment]
After the window is resized, any GetPosition call will give wrong results until the window is moved.

**Comment by flibitijibibo** [comment]
That's new to me... I would throw it on the Bugzilla, but here's our pre-center-window func if you want to look at it:

https://github.com/FNA-XNA/FNA/blob/4cdcfa49d2e36ab060e1e53545c78f0559435458/src/SDL2/SDL2_GameWindow.cs#L238

**Comment by cra0zy** [comment]
Your window centering code is also affected :(

Anyway I opened a bug report: https://bugzilla.libsdl.org/show_bug.cgi?id=3272

**Comment by flibitijibibo** [comment]
If you want to debug it further, my guess is that the resize event needs to poke the window server to get values for the struct; it may be doing something else to assign the position ints. It could be platform-specific, could also be something in a file like SDL_video.c.

**Comment by cra0zy** [comment]
> If you want to debug it further, my guess is that the resize event needs to poke the window server to get values for the struct; it may be doing something else to assign the position ints. It could be platform-specific, could also be something in a file like SDL_video.c.

I did poke it just a bit an hour ago, moved event gets called before resize event and sets the (8, 10) position.

**Comment by cra0zy** [comment]
So for icon loading I decided to use .bmp files because that is the only format available directly from SDL lib, if anyone has any problems with that, speak now.

**Comment by flibitijibibo** [comment]
That's what I did, but you can make PNGs optional if you want:

https://github.com/FNA-XNA/FNA/blob/master/src/SDL2/SDL2_GameWindow.cs#L334

**Comment by cra0zy** [comment]
I'll rebase the SDL branch tomorrow to 1 commit if thats OK with everyone.

**Comment by tomspilman** [comment]
> I'll rebase the SDL branch tomorrow to 1 commit if thats OK with everyone.

I guess that is ok at this point...  lets just not do that too much.

**Comment by cra0zy** [comment]
@flibitijibibo Good news about the resize bug: https://bugzilla.libsdl.org/show_bug.cgi?id=3272

**Comment by tomspilman** [comment]
Ok.... this is scheduled for 3.6 to be released in June.  Lets get this done!

**Comment by cra0zy** [comment]
All the SDL replacement stuff should already be done in the SDL branch, now it needs:
 - custom OpenAL binding
 - custom OpenGL bindings
 - update nuget and project templates

Also I would add:
 - replace System.Drawing code, read from https://github.com/OutOfOrder/MonoKickstart/issues/3#issuecomment-168634616, it would make the dependencies a lot cleaner

**Comment by mrhelmut** [comment]
+1 to get rid of System.Drawing. We wouldn't have to maintain a fork of MonoKickstart anymore.

**Comment by cra0zy** [comment]
> +1 to get rid of System.Drawing. We wouldn't have to maintain a fork of MonoKickstart anymore.

We would still need a fork, but we would only need like 2 libs(SDL and OpenAL).

**Comment by cra0zy** [comment]
@mrhelmut @dellis1972 Has one of you two started working on OpenAL stuff by any chance?

**Comment by dellis1972** [comment]
@cra0zy I've been working on both OpenGL/AL low level stuff. need to find time to update my branch and push the changes.

**Comment by cra0zy** [comment]
@dellis1972 Ok, once you do that I'll help out.

Also you still have admin right over my fork so feel free to add changes directly to SDL2 branch (#4555).

**Comment by matmuze** [comment]
Have any of you guys thought about using bgfx for the graphics backend ?

**Comment by tomspilman** [comment]
@matmuze - Thanks, but we're not looking to wrap another wrapper with our own.

**Comment by cra0zy** [comment]
> @matmuze - Thanks, but we're not looking to wrap another wrapper with our own.

Inception.

**Comment by matmuze** [comment]
Well, I thought that a first step could be to wrap around the wrapper, and then once everything works fine, you guys could replace the function calls from  monogame by the bgfx one. I mean come on they support all the graphics API...it's designed explicitly for what you guys are doing...you must be blind not to see all the benefits you'd get from using such approach right... ?

**Comment by mrhelmut** [comment]
Please read the full thread to understand this design choice.

**Comment by cra0zy** [comment]
> you must be blind not to see all the benefits you'd get from using such approach right... ?

I see 0 benefits:
 - we would have to write a new set of bindings
 - we would need to learn a new API
 - we would have to rewrite our entire API
 - we would probably have to fix some of their code because this is like Xwt or Eto, great on paper, but everything does not work always as expected

**Comment by mrhelmut** [comment]
Big thanks to @cra0zy for #4755 :+1: 

Here are a few things that come to my mind for the next steps (let me know if I'm wrong / incomplete):
- have the template / build targets to deploy SDL2.dll & openal32.dll to build folders (both 32 & 64-bit binaries, with a code to set the custom library path on Windows. see discussion in #4755), and update project template to target Any CPU
- push the new dependencies (inc. the SDL 2.0.5 pre-release / custom patched for Linux)
- replace the AL & GL bindings
- remove OpenTK dependencies from project template and repository
- move to OpenAL Soft
- clean up protobuild definitions to generate DesktopGL projects (it is still named Linux)
- update MonoKickstart dependencies to remove GDI+

I'm going to check if I can get Hi-DPI to work. It's almost there.

**Comment by cra0zy** [comment]
> Big thanks to @cra0zy for #4755

Thanks ^^

> replace the AL & GL bindings
> remove OpenTK dependencies from project template and repository
> move to OpenAL Soft

All of that is getting done in #4555 by @dellis1972  and me, tho I didn't have time to work on it for the last week due to school stuff and me working on some big changes for the Pipeline Tool (Syncing MainWIndow to use Eto.Forms).

If anyone is interested in working on OpenAL side that would be awesome :+1: 

> update MonoKickstart dependencies to remove GDI+

The above work is needed, plus we are still using System.Drawing in some other parts.

> I'm going to check if I can get Hi-DPI to work. It's almost there.

Just a reminder about https://github.com/mono/MonoGame/pull/4755#issuecomment-210530175

**Comment by mrhelmut** [comment]
I'm going to spend some time on this, mostly to get the dll dependencies/targets right because some people using the develop branch seem to be running into the 32/64bit issue.

@cra0zy is there a place where I can find the patched SDL2 for all platforms and cpu? (did @flibitijibibo included your resize fix and the mouse fix in its redist zip?)

I'll also most probably help with the AL wrapper, because I'd like OpenAL-Soft to be used everywhere (and I think that keeping the dll name as soft_oal.dll is better than renaming it openal32.dll, because that would be misleading for people willing to know which implementation is used, and that need the wrapper to be done -- or patch opentk, but let's not talk about this option ^^ --).

**Comment by dellis1972** [comment]
I have a working OpenAL-Soft implementation locally. The OpenAL wrapper is
done its in the SDL2 branch in @cra0zy fork

**Comment by mrhelmut** [comment]
Awesome.

**Comment by cra0zy** [comment]
> (did @flibitijibibo included your resize fix and the mouse fix in its redist zip?)

Yup. tho I think we will still need a 64 bit Windows SDL compiled.

**Comment by mrhelmut** [comment]
Great. Do these fixes allow the window centering code to work with resizable window now, or is this related to another issue?

**Comment by cra0zy** [comment]
> Great. Do these fixes allow the window centering code to work with resizable window now, or is this related to another issue?

It should be fixed in there.

**Comment by mrhelmut** [comment]
How would you like the new runtime dependencies to be handled?

On Windows, SDL.dll & soft_oal.dll need to be present in the output folder. Should they be added to the template? Would it be possible to copy them there without adding them to the template? (so that OS X / Linux builds don't have them in their output folder)

Additionally, the bitness of the dlls should be considered. For that, I thought to output the dlls like this:
- x86/SDL2.dll
- x86/soft_oal.dll
- x64/SDL2.dll
- x64/soft_oal.dll

and then add a code on platform initialization to detect the bitness and call SetDllDirectory (https://msdn.microsoft.com/en-us/library/ms686203(VS.85).aspx).

Any thoughts?

**Comment by cra0zy** [comment]
We need them in output folder for Windows, Linux and Mac. Since the templates are DesktopGL all native libs should get copied out to x86 and x64 folders in the output. This can be achieved with the targets file or by modifying templates.

**Comment by dellis1972** [comment]
the exception being OSX which just needs to be in the root :)

**Comment by mrhelmut** [comment]
I'm not sure to know how to do this with targets, but I'll try to take a look at this.

**Comment by tomspilman** [comment]
> Since the templates are DesktopGL all native libs should get copied out to x86 and x64 folders in the output.

Right... so basically we include all native libs for all platforms/architectures regardless of what platform you're developing on.  It would then be up to the end user to remove the ones unneeded if they want to package or ship for a single platform.

> I'm not sure to know how to do this with targets

You shouldn't need to.

Since we're including all the native binaries for all platforms/architectures then there is no logic involved and .targets are not needed... we just need to copy all of it.  So we can simply include it in the csproj with a "copy if newer" setting.

**Comment by mrhelmut** [comment]
> So we can simply include it in the csproj with a "copy if newer" setting.
This answers my question.

@cra0zy if we output both x86 & x64 libs, is there any mechanism to tell Mono to use one over the other libs? How to make this transparent when debugging on Linux? I don't think SetDllDirectory exist in Mono kernel32 implementation.

**Comment by mrhelmut** [comment]
Would adding an AfterBuild target like this works?

```
  <Target Name="AfterBuild">
	<Copy SourceFiles="$(MonoGameInstallDirectory)\MonoGame\v3.0\Assemblies\DesktopGL\*.dylib" DestinationFolder="$(OutputPath)" ContinueOnError="true" />
	<Copy SourceFiles="$(MonoGameInstallDirectory)\MonoGame\v3.0\Assemblies\DesktopGL\x86\*.*" DestinationFolder="$(OutputPath)x86" ContinueOnError="true" />
	<Copy SourceFiles="$(MonoGameInstallDirectory)\MonoGame\v3.0\Assemblies\DesktopGL\x64\*.*" DestinationFolder="$(OutputPath)x64" ContinueOnError="true" />
  </Target>
```

**Comment by tomspilman** [comment]
> Would adding an AfterBuild target like this works?

I don't think there is any benefit to making it an `<Target Name="AfterBuild">` step.  It just sort of hides things further IMO.

We just need regular items in the project tree:
```
    <None Include="$(MonoGameInstallDirectory)\MonoGame\v3.0\Assemblies\DesktopGL\x86\SDL2.dll">
      <Link>x86\SDL2.dll</Link>
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
```
This then shows up in the project tree make it something the user can delete from his new project if he likes.

Or is that a bad feature to have?

**Comment by hach-que** [comment]
See
https://protobuild.readthedocs.io/en/latest/external_projects.html#referencing-native-binaries-on-disk
on how to do that.

**Comment by hach-que** [comment]
You should declare native binaries in the Protobuild definitions instead.

**Comment by tomspilman** [comment]
> You should declare native binaries in the Protobuild definitions instead.

There is no Protobuild definition for this yet... we're talking about Visual Studio templates.

**Comment by cra0zy** [comment]
@mrhelmut Not sure... will have to test it.

**Comment by hach-que** [comment]
Aren't these also new dependencies of MonoGame itself though? So anyone using MonoGame in source form will need NativeBinary declared for things to work right (automatically).

**Comment by flibitijibibo** [comment]
I don't know how deep VS can check for platform target/arch, but what I'd recommend is just copying the binaries needed to make the game boot for the developer, and then make a packaging target that puts everything in the right place for when they _actually_ put together app bundles, MojoSetup packages, etc. for people to install.

When getting a game to run, nobody cares about the other targets in day-to-day development, particularly if the platform code is reliable enough for them to say "eh, I can just deploy and test without the debugger or whatever." But, they'll definitely want to know what needs to go where when it's finally time to ship to customers.

I think it's reasonable to tell devs not to dump their Release/ folder into SteamPipe and call it a day. Having a process like this (automated or not) is something that should be considered for actually shipping vs. just compiling the game:

https://github.com/FNA-XNA/FNA/wiki/3:-Distributing-FNA-Games

**Comment by mrhelmut** [comment]
I agree, and this reminds me that I should continue #4430.

**Comment by cra0zy** [comment]
and that reminds me that I should continue https://github.com/MonoGame/MonoGame.Packager

I kind of put that on hold till this was done.

**Comment by hach-que** [comment]
At some point in the future I intend on having a process for packaging Protobuild projects for redistribution (like --pack, but one that creates a package for a target distribution platform, rather than a package for the package index).  This is likely a while away though.

However, redistribution is separate from the discussion happening here.  I think without the changes being discussed, even running it from the debugger won't work because the files aren't copied across in the project / visual studio template?

**Comment by mrhelmut** [comment]
>However, redistribution is separate from the discussion happening here. I think without the changes being discussed, even running it from the debugger won't work because the files aren't copied across in the project / visual studio template?

You're right. What's important is that debugging the template works as-is on any platform/arch.

**Comment by hach-que** [comment]
Until the new project generator used, the right way forward for the Visual Studio templates is @tomspilman's suggestion to add `<None>` copy-on-build to the project.  This is effectively what "NativeBinary" does under the hood in Protobuild (but with a bit more smarts).

We still need "NativeBinary" added to the MonoGame third party dependencies though, so that people using MonoGame in source mode will have things work.  I'm not sure if that's already done (this issue itself isn't a pull request, so I'm not sure the context of these discussions).

**Comment by mrhelmut** [comment]
I'll make sure to update both the template and MonoGame project definition.

**Comment by KonajuGames** [comment]
The dllconfig files used by Mono support an architecture field so they can
load different libraries for different architectures. That is what you will
need to use here.

**Comment by hach-que** [comment]
@KonajuGames Those don't work on Windows under .NET though.

What I do for native architectures is have the DLLs named like NativeDLL_x86.dll and NativeDLL_x64.dll, and when the application starts up I copy the DLL to NativeDLL.dll depending on the architecture of the application as it's running.

**Comment by flibitijibibo** [comment]
>I copy the DLL to NativeDLL.dll

Huh, does that actually work on Windows nowadays? I remember getting some antivirus warnings from FEZ Windows players when Renaud did that for OpenAL Soft at launch...

**Comment by cra0zy** [comment]
@KonajuGames @hach-que Both of you are missing the problem.

On Windows we have "SetDllDirectory" so we can have native libs in x86 and x64 folders.
On Mac there is only 1 set of libraries so it's gonna be placed directly.

The question is does "SetDllDirectory" work on LInux which I will test in 2 mins.

**Comment by mrhelmut** [comment]
There's already a MonoGame.Framework.dll.config ready, I totally forgot about it.

I'm really keen to copy dll at runtime, I'll try to detect the bitness at runtime and set the path with SetDllDirectory from kernel32.dll. Overriding the PATH env var would work too I think.

**Comment by cra0zy** [comment]
@flibitijibibo Did you disable Wayland support when you compiled Linux SDL for FNA?

**Comment by flibitijibibo** [comment]
@cra0zy Nah, my buildbots don't have the Wayland/Mir headers/libs so they get skipped in the build. I'll probably fix that once they start shipping For Real(TM) but for now I'm depending on the DynAPI in SDL2 to swap out my VideoInit with one that supports those video drivers:

https://hg.libsdl.org/SDL/file/7cbfd97f1430/docs/README-dynapi.md

**Comment by hach-que** [comment]
@flibitijibibo *shrugs* I have no idea what anti-virus nonsense vendors do these days.  I just have the built-in Windows Defender and it works fine.

**Comment by KonajuGames** [comment]
For Windows, use SetDllDirectory.
For Mac, use the dll.config to point to the fat binary with both architectures in the one dylib.
For Linux, have two entries in the dll.config with the `cpu` field as `x86` for one and `x86-64` for the other, as demonstrated [here](http://www.mono-project.com/docs/advanced/pinvoke/dllmap/) and described [here](http://www.mono-project.com/docs/advanced/pinvoke/dllmap/).

**Comment by cra0zy** [comment]
http://mono.1490590.n4.nabble.com/DllImport-and-modifying-library-search-path-td3300872.html

Cpu can be specified in the .config file... good to know.

**Comment by cra0zy** [comment]
You posted moments before me O o.

**Comment by mrhelmut** [comment]
Awesome, I'll update the file and SetDllDirectory for Windows. :+1: 
Just waiting for all the dependencies (MonoGame/MonoGame.Dependencies#84 + @cra0zy Linux binaries) to be merged and I already have a branch with the template and the Windows installer update ready to be pushed.

I'm not confident enough to update the OS X and Linux installer.

**Comment by KonajuGames** [comment]
The dependencies are now merged.

**Comment by mrhelmut** [comment]
Regarding MonoGame.Framework.dll.config...

If I add paths to the dllmap for the each cpu arch like this:
```
<dllmap dll="SDL2.dll" os="osx" target="libSDL2.dylib" />
<dllmap dll="SDL2.dll" os="linux" cpu="x86" target="./x86/libSDL2-2.0.so.0" />
<dllmap dll="SDL2.dll" os="linux" cpu="x64_86" target="./x64/libSDL2-2.0.so.0" />
```
it won't be compatible with MonoKickstart (which sets the dll path in its bootstrap script) and would require a different .dll.config (or alternatively, a double folder architecture like "./lib64/x64/"...).

The other way around would be to have all native lib at the root of the output folder and rename them. I'm not very keen about this, because Windows would still require separate folders and would look misleading.

Are we ok with the .dll.config not being compatible as-is with MonoKickstart? I think that the most important thing is that it's as transparent as possible to debug projects, and that anything related to packaging / publishing can be handled through tools or documentation.

**Comment by mrhelmut** [comment]
I have a branch ready to be added to #4874 here: https://github.com/mrhelmut/MonoGame/tree/SDLDep

It adds:
- All dependencies to MonoGame definitions
- All dependencies to the Windows installer
- Updates the DesktopGL VS2010 template
- Rename "openal32.dll" to "soft_oal.dll" in order to reflect the switch to OpenAL-Soft
- Add code to set the DLL directory on runtime for Windows (https://github.com/mrhelmut/MonoGame/commit/65164e33e8ebc88b13c9c32f0b1cd48cf48d9afb#diff-46e6b4145d3c3fe343f455f6aeab9538R87)

What it doesn't do:
- Update Linux installer / template
- Update OS X installer / template

@cra0zy @dellis1972

**Comment by mrhelmut** [comment]
The Windows installer and the template work just fine, as well as the DLL loading (test PR #4883, if you'd like to test the artifacts).

All lights are green on Windows for me, awesome work. :)

@cra0zy would like me to push https://github.com/mrhelmut/MonoGame/tree/SDLDep commits to your NewSDL branch?


(Mind that the new template has 2 Icon files -- BMP + ICO --, this is because the current implementation can't load .ico for the SDL window title, and still need an .ico for the exe icon. This will be fixed once we switch to another ico loading method.)

**Comment by mrhelmut** [comment]
I squashed all the template/installer update into this commit: https://github.com/mrhelmut/MonoGame/commit/65b8481868c69ede5e55a4688e5ea96d87dac500

**Comment by mrhelmut** [comment]
@tomspilman This is fixed. :)

We should maybe schedule a few maintenance operations on the repo' (e.g. MonoMac, renaming Linux definition to DesktopGL, removing WindowsGL...).

**Comment by cra0zy** [comment]
No, it isn't, there are still Android and iOS.

**Comment by mrhelmut** [comment]
Oh, my bad!

**Comment by dellis1972** [comment]
I'm working on Android and iOS the first bit is done, need to fix the
GameViews etc now though.

**Comment by cra0zy** [comment]
SDL 2.0.5 will be released soon: http://www.phoronix.com/scan.php?page=news_item&px=SDL-2.0.5-Prepping

@flibitijibibo Did you submit your patch to allow changing resizable state of the window upstream? / Was it merged?

**Comment by flibitijibibo** [comment]
Everything except the VS debugger fix is in, I'm fixing that now.

**Comment by cra0zy** [comment]
Sweet.

**Comment by mrhelmut** [comment]
Awesome news.

**Comment by flibitijibibo** [comment]
Right, so here's how that VS fix is going to work:

Because the exception is only a problem in a very specific case, and there's no real way to specifically detect .NET debuggers, we're going to use an SDL_Hint to sidestep the issue. Here's what the SDL-mirror diff looks like...

https://github.com/FNA-XNA/SDL-mirror/commit/2d06bf659acd2bc97a1fc80fe078d178b8f566f3

... and this is what we have to write in:

https://github.com/FNA-XNA/FNA/commit/1b38ac36ed0ea86a326235e03ab8bb63100822d6

As long as it's before SDL_Init you should be fine. And if that hint name changes before we mainline it, I'll let you know.

**Comment by flibitijibibo** [comment]
SDL 2.0.5 is out: http://lists.libsdl.org/pipermail/sdl-libsdl.org/2016-October/100786.html

All the critical changes for MG are already in fnalibs, but I'll be updating to 2.0.5 mainline when FNA 16.11 is out.

**Comment by tomspilman** [comment]
@dellis1972 - So is this done now?  Is all of OpenTK gone?

**Comment by cra0zy** [comment]
All gone.

**Comment by tomspilman** [comment]
Thanks @cra0zy !

**Comment by tomspilman** [comment]
Well actually....

https://github.com/MonoGame/MonoGame/search?p=1&q=OpenTK&type=&utf8=%E2%9C%93

We should purge out these last remaining reference to OpenTK if we are indeed not using it.

**Comment by dellis1972** [comment]
the only platform using OpenTK now is the Native Mac platform. But since the templates for Xamarin.Mac all use DesktopGL assemblies now I'm not sure if we need it around. The rest seems to be comment clean up.

**Comment by dellis1972** [comment]
We can probably push this to 3.8 if needed.

**Comment by tomspilman** [comment]
@dellis1972 - Could you write a new issue for this in 3.8 listing whatever work is left?

**Comment by dellis1972** [comment]
will do
